Add 3 example use-cases to wildfly-camel-examples

jochenr opened the pull request:
I'd like to conribute my example uses-cases to the "official" wildfly-camel-examples, because I hope that they will help others using the camel integration on wildfly.

My use-cases are:

1. camel-cxf-jaxws-cdi-xml
This example uses a XML route definition while the endpoints and processors are produced with CDI.
For us this example is important for the migration of our current Switchyard applications to wildfly-camel.

2. camel-cxf-jaxws-security
This example shows how to use http-basic authentication.
It uses XML-DSL route builder with spring configuration.

3. camel-cxf-jaxws-security-cdi
This example shows how to use client-cert-/tls-mutual-authentication.
It uses Java-DSL route builder with CDI
The test "org.wildfly.camel.test.cxf.CxfWsSecurityCdiExampleTest" will work after https://github.com/wildfly-extras/wildfly-camel/issues/1945 is fixed

jamesnetherton commented on camel-cxf-jaxws-cdi-xml/pom.xml:
```diff
@@ -0,0 +1,120 @@
+        <version>4.9.0-SNAPSHOT</version>
+    </parent>
+
+    <name>Wildfly Camel :: Example :: Camel CXF JAX-WS CXF CDI XML-Configuration</name>
```
Maybe just: `Wildfly Camel :: Example :: Camel CXF JAX-WS CDI XML`

jamesnetherton commented on camel-cxf-jaxws-cdi-xml/pom.xml:
```diff
@@ -0,0 +1,120 @@
+        <!--
+        <dependency>
+            <groupId>org.wildfly.camel</groupId>
+            <artifactId>wildfly-camel-subsystem-core</artifactId>
```
Please can you remove any stuff that's commented out and not needed.

jamesnetherton commented on camel-cxf-jaxws-cdi-xml/pom.xml:
```diff
@@ -0,0 +1,120 @@
+
+		<dependency>
+			<groupId>org.apache.camel</groupId>
+			<artifactId>camel-cxf</artifactId>
```
There's a few things that aren't formatted nicely. There are some IDE code style templates which may help over at the main project:

**Eclipse:**

https://github.com/wildfly-extras/wildfly-camel/tree/master/docs/eclipse

**Intellij:**

https://github.com/wildfly-extras/wildfly-camel/tree/master/docs/intellij

jamesnetherton commented on camel-cxf-jaxws-cdi-xml/src/main/java/org/wildfly/camel/examples/cxf/jaxws/Application.java:
```diff
@@ -0,0 +1,73 @@
+	public CxfEndpoint produceCxfConsumer() {
+		CxfComponent cxfComponent = new CxfComponent(this.context);
+		CxfEndpoint cxfFromEndpoint = new CxfEndpoint("http://localhost:8080/webservices/greeting-cdi-xml", cxfComponent);
+		cxfFromEndpoint.setServiceClass(org.wildfly.camel.examples.cxf.jaxws.GreetingService.class);
```
Do we need the fully qualified class name here or can we just simplify with `GreetingService.class`?

jamesnetherton commented on camel-cxf-jaxws-cdi-xml/src/main/resources/cxfws-cdi-xml-camel-context.xml:
```diff
@@ -0,0 +1,43 @@
+         CamelCxfWsServlet sends the required message and name parameters to the service by populating them
+         within the message exchange body.
+        -->
+        <route id="_route1">
```
Please can we drop the underscore(`_`) from element IDs.

jochenr replied:
I'll fix it.
Unfortunatley this is a default behaviour of the latest Red Hat Developer Studio Editor for Camel.
It changes things, when toggling bewteen "Design" and "Source" view;-(

jamesnetherton commented on camel-cxf-jaxws-cdi-xml/src/main/webapp/WEB-INF/jboss-web.xml:
```diff
@@ -0,0 +1,24 @@
+  -->
+
+<jboss-web>
+    <context-root>example-camel-cxfws-cdi-xml</context-root>
```
There's been a slight change in the latest version of the examples codebase. I removed all of the `jboss-web.xml` from each example project and have simplified things by setting `finalName` in the Maven `pom.xml`. See here for an example:

https://github.com/wildfly-extras/wildfly-camel-examples/blob/master/camel-cdi/pom.xml#L64

jamesnetherton commented on camel-cxf-jaxws-security/pom.xml:
```diff
@@ -0,0 +1,112 @@
+        <version>4.9.0-SNAPSHOT</version>
+    </parent>
+
+    <name>Wildfly Camel :: Example :: Camel CXF JAX-WS with Security</name>
```
Like the other secure project lets rename to `camel-secure-cxf-jaxws`. And have the name description as:

`Wildfly Camel :: Example :: Camel Secure CXF JAX-WS`

jamesnetherton commented on camel-cxf-jaxws-security/src/main/webapp/WEB-INF/cxfws-security-camel-context.xml:
```diff
@@ -0,0 +1,75 @@
+         CamelCxfWsServlet sends the required message and name parameters to the service by populating them
+         within the message exchange body.
+        -->
+        <route id="_route1">
```
Drop the underscores.

jamesnetherton commented on itests/pom.xml:
```diff
@@ -186,6 +204,14 @@
                 </excludes>
             </resource>
+			<resource>				
+				<directory>src/main/resources/configuration</directory><!-- from -->
```
Please remove the comment

jamesnetherton commented on itests/pom.xml:
```diff
@@ -186,6 +204,14 @@
             </resource>
+			<resource>				
+				<directory>src/main/resources/configuration</directory><!-- from -->
+				<targetPath>${jboss.home}/standalone/configuration</targetPath><!-- to -->
```
As above, no comment required here.

jamesnetherton commented on itests/pom.xml:
```diff
@@ -262,8 +288,67 @@
+                            <environmentVariables>
+                                <JBOSS_HOME>${jboss.home}</JBOSS_HOME>
+                            </environmentVariables>
+                            <commandlineArgs>-jar ${jboss.home}/jboss-modules.jar -mp ${jboss.home}/modules org.jboss.as.domain-add-user -a -s -u testUser -p testPassword1+ -g testRole</commandlineArgs>
```
Is there a way for this to be done within the test class at runtime via the DMR?

If possible, I'd prefer to keep this sort of thing out of the Maven build. The example tests are shared in other projects which means we'd have to port over and duplicate the additional maven executions.

jochenr replied:
I'll give it a try, but I'm not sure if this is possible. I thought DMR can only execute CLI scripts to "manipulate" standalone-xxx.xml configuration.
Can I really add a user with it at runtime?

jamesnetherton replied:
Maybe it's not possible at runtime. I'll think about this some more

Perhaps manipulating the user.properties files at runtime is good enough? I'm not sure.

jochenr replied:
I have an idea...please see my commit in some minutes....

jamesnetherton commented on itests/pom.xml:
```diff
@@ -262,8 +288,67 @@
+                            <environmentVariables>
+                                <JBOSS_HOME>${jboss.home}</JBOSS_HOME>
+                            </environmentVariables>
+                            <commandlineArgs>-jar ${jboss.home}/jboss-modules.jar -mp ${jboss.home}/modules org.jboss.as.domain-add-user -a -s -u CN=localhost -p testPassword1+ -g testRole</commandlineArgs>
```
Previous comment applies here too.

jamesnetherton commented on itests/src/main/java/org/wildfly/camel/test/cxf/CxfWsSecurityCdiExampleTest.java:
```diff
@@ -0,0 +1,62 @@
+
+@RunAsClient
+@RunWith(Arquillian.class)
+public class CxfWsSecurityCdiExampleTest {
```
Rename to `CxfSecureWsCdiExampleTest`.

jamesnetherton commented on itests/src/main/java/org/wildfly/camel/test/cxf/CxfWsSecurityExampleTest.java:
```diff
@@ -0,0 +1,56 @@
+
+@RunAsClient
+@RunWith(Arquillian.class)
+public class CxfWsSecurityExampleTest {
```
Rename to `CxfSecureWsExampleTest`.

jochenr commented:
should I also move the CLI for creating the login-modules and system properties from the maven build into the test class?

jamesnetherton commented:
Ideally, yes. I'd prefer to not have it in the `pom.xml`.

jochenr commented:
Please review my last commit especially the test class "CxfSecureWsCdiExampleTest".
I moved the CLI commands from pom.xml to that class, but I have difficulties, because there is no blocking "reload" command. Thats why I had to put an ugly Thread.sleep(...) before the test can be run.
Unfortunately I didn't manage to get Arquillian to do the restart. Perhaps you can help with that, since I don't have a lot of experience with Arquillian.
Thanks

jamesnetherton commented:
Thanks!

I have some of my own minor tweaks to apply on top of this work, so if it's all good, I'll merge to master  (probably later in the week).

For now, I'm probably going to drop `camel-cxf-jaxws-security-cdi` until we have a proper resolution to the issues you raised over at the main wildfly-camel project. I'd prefer to not promote hacks and workarounds if possible. We can add it in once we have a better solution.

jochenr commented:
Sounds good, thank you.
I hope I'll have a good idea when #1945 is fixed.

jochenr commented:
The solution for the :reload problem after the execution of the CLI scripts can be found here:
https://github.com/wildfly/wildfly/blob/master/testsuite/shared/src/main/java/org/jboss/as/test/shared/ServerReload.java

Used in this example:
https://github.com/wildfly/wildfly/blob/master/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/bootstrap/CustomBootstrapContextTestCase.java#L102
https://github.com/wildfly/wildfly/blob/master/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/JcaMgmtServerSetupTask.java
https://github.com/wildfly/wildfly/blob/master/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/JcaMgmtBase.java

as solution for this issue:
https://issues.jboss.org/browse/WFLY-91

I'll do a solution based on this official Wildfly testcase after my return from my (long) holiday.

Should I put a dependency to <artifactId>wildfly-testsuite-shared</artifactId> in the itests pom or should I copy the class "org.jboss.as.test.shared.ServerReload"?

jamesnetherton commented:
My preference would be to copy the class (keeping the license attribution), as IIRC, the testsuite-shared dependency pulls in all kinds of extraneous junk.

Keep in mind, I'm not going to merge the `camel-cxf-jaxws-security-cdi` code until we have fixes for the other CXF issues. So no need to push these changes onto this PR.

I'll try and merge the others later today.

jamesnetherton commented:
@jochenr Does `camel-cxf-jaxws-cdi-xml` work for you? I can't get it to deploy without errors.

Maybe I changed something locally that broke it, but everything looks valid....

jochenr commented:
yes, it works for me.
I only need a workaround (decribed in #1945) for camel-cxf-jaxws-security-cdi

what Exception do you get at deploy time?

jamesnetherton commented:
`@ImportResource` seems to be having no affect. There are no messages in the server logs to say that the Camel Context has been started.

If it works ok for you, I'll assume I've changed and broken something locally. I'll dig into it some more.

jochenr commented:
Did you do the same mistake as I did (and opened https://github.com/wildfly-extras/wildfly-camel/issues/1928) ?
Is the xml file from @ImportResource copied in the classpath of the war? I think I changed the resources part of the pom to get it included...

P.S.: I just changed my profile so that you can see my (private) e-mail adress...if you like to send me a stacktrace...

jamesnetherton commented:
Yep - I screwed up. My IDE had modified some of the bean ids in the Camel Context XML.

jamesnetherton commented:
camel-cxf-jaxws-cdi-xml and camel-cxf-jaxws-security have been merged to master with some minor mods.

The CDI security example will have to wait, pending the fixes to the CXF issues you raised over at the main project.

Thanks for contributing :smile:
